thrust/system/cuda/detail/bulk/malloc.hpp

```diff
@@ -38,9 +38,24 @@ inline __device__ bool is_on_chip(void *ptr)
 template<typename T>
 inline __device__ T *on_chip_cast(T *ptr)
 {
+#ifndef __clang__
```

Comment:
I think I would make the UB code depend on NVCC rather than !clang. The desired effect seems to depend on an optimization that NVCC applies.